### PR TITLE
Introduce GT_PROMISE for __builtin_assume

### PR DESCRIPTION
**Pull request by havogt**
Credits to @lukasm91 for hinting to `__builtin_assume` in `deref`. 

On more recent compilers (than 11.2 what we have in CI on daint), will improve codegen to the level of #1779, but is save.

### CODE REVIEW

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch perftests

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
TODO switch to CUDA 11.2 on daint for testing

File: include/gridtools/common/defs.hpp [hunk 1]
```diff
@@ -26,6 +27,21 @@ namespace gridtools {
 #endif
 #endif
 
+#if defined(__CUDACC__) &&                                                                       \
+    (__CUDACC_VER_MAJOR__ >= 12 || (__CUDACC_VER_MAJOR__ == 11 && __CUDACC_VER_MINOR__ >= 2)) && \
+    defined(__CUDA_ARCH__)
```
Comment (fthaler) [inline include/gridtools/common/defs.hpp:32]: Maybe this is enough?
```suggestion
#if defined(__has_builtin) && __has_builtin(__builtin_assume)
```

Comment (havogt) [reply include/gridtools/common/defs.hpp:32]: I like it.

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch perftests

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch perftests

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
As expected no perf difference on the old compiler.

File: include/gridtools/fn/unstructured.hpp [hunk 2]
```diff
@@ -18,6 +18,7 @@
 #include "./common_interface.hpp"
 #include "./executor.hpp"
 #include "./neighbor_table.hpp"
+#include "gridtools/common/defs.hpp"
```
Comment (havogt) [inline include/gridtools/fn/unstructured.hpp:21]:
```suggestion
#include "../common/defs.hpp"
```

**Comment by havogt** [comment]
launch jenkins

**Review by fthaler — APPROVED** [review]

**Review by iomaganaris — APPROVED** [review]

**Pull request merged**